scipy/optimize/_zeros_py.py

```diff
@@ -1392,3 +1393,344 @@ def toms748(f, a, b, args=(), k=1,
+        # For now, we assume that `func` requires arguments of the original
+        # shapes. However, `x` is compressed (contains only active elements).
+        # Therefore, we inflate `x` by creating a full-size array, copying the
+        # elements of `x` into it, and making it the expected shape.
+        # TODO: allow user to specify that `func` works with compressed input
+        x_full = res.x.copy()
+        x_full[active] = x
+        x_full = x_full.reshape(shape)
+        f = func(x_full, *args)
+        nfev += 1
+        # Ensure that the outpuf of `func` is an array of the appropriate
+        # dtype, ravel it (because we work with 1D arrays for simplicity), and
+        # compress it to contain only active elements.
+        f = np.asarray(f, dtype=dtype).ravel()[active]
+
+        # [1] Flowchart 2 (flowchart is reversed; see code)
+        x3, f3 = x2.copy(), f2.copy()
+        j = np.sign(f) == np.sign(f1)
+        nj = ~j
+        x3[j], f3[j] = x1[j], f1[j]
+        x2[nj], f2[nj] = x1[nj], f1[nj]
+        x1, f1 = x, f
+
+        # [1] Flowchart 3
+        nit += 1
+        temp = _check_termination(x1, f1, x2, f2, x3, f3,
+                                  res, active, status, nfev, nit, tols)
+        xmin, fmin, x1, f1, x2, f2, x3, f3, active, status, tl = temp
+
+        if callback is not None:
+            temp_res = _prepare_result(shape, res, active, nit, nfev)
+            if _call_callback_maybe_halt(callback, temp_res):
+                cb_terminate = True
+                break
+        if active.size==0:
+            break
+
+        # [1] Flowchart 4 / Equation 1
+        xi1 = (x1 - x2) / (x3 - x2)
+        phi1 = (f1 - f2) / (f3 - f2)
+        alpha = (x3 - x1) / (x2 - x1)
+        j = ((1 - np.sqrt(1 - xi1)) < phi1) & (phi1 < np.sqrt(xi1))
+
+        f1j, f2j, f3j, alphaj = f1[j], f2[j], f3[j], alpha[j]
+        t = np.full_like(alpha, 0.5)
+        t[j] = (f1j / (f1j - f2j) * f3j / (f3j - f2j)
+                - alphaj * f1j / (f3j - f1j) * f2j / (f2j - f3j))
+
+        # See [1] Appendix, "Adjust T Away from the Interval Boundary"
+        t = np.clip(t, tl, 1-tl)
+
+    res.status[active] = _ECALLBACK if cb_terminate else _ECONVERR
+    return _prepare_result(shape, res, active, nit, nfev)
+
+def _chandrupatla_iv(func, a, b, args, xatol, xrtol,
+                     fatol, frtol, maxiter, callback):
+
+    if not callable(func):
+        raise ValueError('`func` must be callable.')
+
+    # a and b have more complex IV that is taken care of during initialization
+
+    if not np.iterable(args):
+        args = (args,)
+
+    # rudimentary, but enough IMO
+    tols = np.asarray([xatol, xrtol, fatol if fatol is not None else 1, frtol])
+    if (not np.issubdtype(tols.dtype, np.number)
+            or np.any(tols < 0)
+            or tols.shape != (4,)):
+        raise ValueError('Tolerances must be non-negative scalars.')
+
+    maxiter_int = int(maxiter)
+    if maxiter != maxiter_int or maxiter < 0:
```

Comment:
What I did above with `tols` was a compact way of doing all four tolerances in one shot. We could make this the same for consistency, but it would be an unusual way of doing input validation of an integer if taken out of context.